hspec-discover/src/Run.hs

```diff
@@ -119,15 +119,31 @@ findSpecs src = do
   mapMaybe (fileToSpec dir) . filter (/= file) <$> getFilesRecursive dir
 
 fileToSpec :: FilePath -> FilePath -> Maybe Spec
-fileToSpec dir file = case reverse $ splitDirectories file of
-  x:xs -> case stripSuffix "Spec.hs" x <|> stripSuffix "Spec.lhs" x of
-    Just name | (not . null) name -> Just . Spec (dir </> file) $ (intercalate "." . reverse) (name : xs)
```

Comment:
We could just replace `(not . null) name` with `isValidModuleName name && all isValidModuleName xs`.  I would prefer that not only because it's a smaller change, but also for the reasons given in my other comments.